File: latest/week2/workshop5_m5/W05_M05_demo.py
"""W05-M05: Introduction to Spatial Data — Python — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
os.makedirs("output", exist_ok=True)

# 1. CREATE SPATIAL DATA FROM SCRATCH
cities = pd.DataFrame({
    "name": ["Tirana", "Durrës", "Vlorë", "Shkodër", "Elbasan", "Korçë"],
    "lon": [19.82, 19.45, 19.49, 19.51, 20.08, 20.78],
    "lat": [41.33, 41.32, 40.47, 42.07, 41.11, 40.62],
    "pop_2023": [420000, 115000, 79000, 77000, 78000, 51000]})

# Convert to GeoDataFrame
try:
    import geopandas as gpd
    from shapely.geometry import Point, Polygon

    # Points from lat/lon columns
    geometry = [Point(lon, lat) for lon, lat in zip(cities.lon, cities.lat)]
    cities_gdf = gpd.GeoDataFrame(cities, geometry=geometry, crs="EPSG:4326")
    print("GeoDataFrame created:")
    print(cities_gdf)
    print(f"CRS: {cities_gdf.crs}")

    # 2. PLOT
    fig, ax = plt.subplots(figsize=(6, 7))
    cities_gdf.plot(ax=ax, markersize=cities_gdf["pop_2023"] / 5000,
        color="#E53935", alpha=0.6, edgecolor="white", lw=0.5)
    for _, row in cities_gdf.iterrows():
        ax.text(row.geometry.x + 0.05, row.geometry.y + 0.05,
            row["name"], fontsize=7, fontweight="bold")
    ax.set_title("Albanian Cities (GeoDataFrame, EPSG:4326)", fontweight="bold")
    ax.set_xlabel("Longitude"); ax.set_ylabel("Latitude")
    ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
    plt.tight_layout(); plt.savefig("output/cities_gpd.png", dpi=300); plt.close()

    # 3. CRS TRANSFORMATION
    cities_utm = cities_gdf.to_crs(epsg=32634)
    print(f"\nUTM CRS: {cities_utm.crs}")

    # 4. SIMPLIFIED ALBANIA POLYGON
    alb_coords = [(19.3,39.6),(19.8,39.7),(20.1,39.8),(20.6,40.0),
        (21.0,40.5),(20.7,41.0),(20.5,41.5),(20.3,42.0),
        (20.0,42.1),(19.5,41.8),(19.3,41.5),(19.0,40.5),(19.3,39.6)]
    alb_poly = gpd.GeoDataFrame({"name": ["Albania"]},
        geometry=[Polygon(alb_coords)], crs="EPSG:4326")

    # Overlay
    fig, ax = plt.subplots(figsize=(6, 7))
    alb_poly.plot(ax=ax, facecolor="#E3F2FD", edgecolor="#1565C0", linewidth=1.5)
    cities_gdf.plot(ax=ax, markersize=cities_gdf["pop_2023"] / 5000,
        color="#E53935", alpha=0.6, edgecolor="white", lw=0.5, zorder=5)
    for _, row in cities_gdf.iterrows():
        ax.text(row.geometry.x + 0.05, row.geometry.y + 0.05,
            row["name"], fontsize=7, fontweight="bold")
    ax.set_title("Albania: Points on Polygon", fontweight="bold")
    ax.set_aspect("equal"); ax.axis("off")
    plt.tight_layout(); plt.savefig("output/overlay_gpd.png", dpi=300); plt.close()

    # 5. FILE I/O
    cities_gdf.to_file("output/cities.geojson", driver="GeoJSON")
    cities_gdf.to_file("output/cities.gpkg", driver="GPKG")
    back = gpd.read_file("output/cities.geojson")
    print(f"\nRead back: {len(back)} features")

except ImportError:
    print("geopandas not installed — using matplotlib fallback")
    fig, ax = plt.subplots(figsize=(6, 7))
    # Simple Albania outline
    lons = [19.3,19.8,20.1,20.6,21.0,20.7,20.5,20.3,20.0,19.5,19.3,19.0,19.3]
    lats = [39.6,39.7,39.8,40.0,40.5,41.0,41.5,42.0,42.1,41.8,41.5,40.5,39.6]
    ax.fill(lons, lats, alpha=0.15, color="#1565C0", edgecolor="#1565C0", lw=2)
    ax.scatter(cities.lon, cities.lat, s=cities.pop_2023/5000,
        c="#E53935", alpha=0.6, edgecolors="white", lw=0.5, zorder=5)
    for _, row in cities.iterrows():
        ax.text(row.lon + 0.05, row.lat + 0.05, row["name"], fontsize=7, fontweight="bold")
    ax.set_title("Albania (matplotlib fallback)", fontweight="bold")
    ax.set_aspect("equal")
    plt.tight_layout(); plt.savefig("output/albania_fallback.png", dpi=300); plt.close()

# 6. FOLIUM INTERACTIVE (standalone HTML)
try:
    import folium
    m = folium.Map(location=[41.33, 19.82], zoom_start=7,
        tiles="CartoDB positron")
    for _, row in cities.iterrows():
        folium.CircleMarker(
            location=[row["lat"], row["lon"]],
            radius=row["pop_2023"] ** 0.5 / 100,
            color="#E53935", fill=True, fill_opacity=0.6,
            popup=f"<b>{row['name']}</b><br>Pop: {row['pop_2023']:,}",
            tooltip=row["name"]
        ).add_to(m)
    m.save("output/albania_cities.html")
    print("Interactive map saved: output/albania_cities.html")
except ImportError:
    print("folium not installed — skipping interactive map")

print("\nAll W05-M05 Python outputs saved")

File: latest/week2/workshop5_m5/W05_M05_demo.R
# ============================================================
# Workshop 5 — Module 5: Introduction to Spatial Data
# Integrated R Script — UNYT Tirana
# ============================================================
library(tidyverse); library(sf); dir.create("output", showWarnings = FALSE)

# ── 1. CREATE SPATIAL DATA FROM SCRATCH ──────────────────
# Albanian cities as a tibble with coordinates
cities <- tibble(
  name     = c("Tirana", "Durrës", "Vlorë", "Shkodër", "Elbasan", "Korçë"),
  lon      = c(19.82, 19.45, 19.49, 19.51, 20.08, 20.78),
  lat      = c(41.33, 41.32, 40.47, 42.07, 41.11, 40.62),
  pop_2023 = c(420000, 115000, 79000, 77000, 78000, 51000))

# Convert to sf POINT object (CRS = WGS 84)
cities_sf <- st_as_sf(cities, coords = c("lon", "lat"), crs = 4326)
cat("Class:", class(cities_sf), "\n")
cat("CRS:", st_crs(cities_sf)$input, "\n")
print(cities_sf)

# ── 2. PLOT WITH ggplot2 + geom_sf ──────────────────────
p_cities <- ggplot(cities_sf) +
  geom_sf(aes(size = pop_2023), color = "#E53935", alpha = 0.6) +
  geom_sf_text(aes(label = name), size = 3, nudge_y = 0.08) +
  scale_size(range = c(2, 12), name = "Population") +
  theme_void() +
  labs(title = "Albanian Cities (sf POINT, EPSG:4326)")
ggsave("output/cities_sf.png", p_cities, width = 6, height = 7, dpi = 300)

# ── 3. CRS TRANSFORMATION ───────────────────────────────
# Transform to UTM Zone 34N for distance calculations
cities_utm <- st_transform(cities_sf, crs = 32634)
cat("\nUTM CRS:", st_crs(cities_utm)$input, "\n")

# Distance matrix (in metres)
dist_mat <- st_distance(cities_utm)
rownames(dist_mat) <- cities$name; colnames(dist_mat) <- cities$name
cat("\nDistance Tirana–Vlorë:", round(dist_mat["Tirana", "Vlorë"] / 1000, 1), "km\n")
cat("Distance Tirana–Shkodër:", round(dist_mat["Tirana", "Shkodër"] / 1000, 1), "km\n")

# ── 4. READ SPATIAL FILE (GeoJSON) ──────────────────────
# If you have a GeoJSON file:
# alb <- st_read("albania_districts.geojson")
# st_crs(alb)              # check CRS
# ggplot(alb) + geom_sf()  # quick plot

# Simulated polygon (simplified Albania outline)
alb_coords <- matrix(c(
  19.3, 39.6,  19.8, 39.7,  20.1, 39.8,  20.6, 40.0,
  21.0, 40.5,  20.7, 41.0,  20.5, 41.5,  20.3, 42.0,
  20.0, 42.1,  19.5, 41.8,  19.3, 41.5,  19.0, 40.5,
  19.3, 39.6), ncol = 2, byrow = TRUE)
alb_poly <- st_sf(name = "Albania",
  geometry = st_sfc(st_polygon(list(alb_coords)), crs = 4326))

# Overlay cities on country
p_overlay <- ggplot() +
  geom_sf(data = alb_poly, fill = "#E3F2FD", color = "#1565C0", linewidth = 1) +
  geom_sf(data = cities_sf, aes(size = pop_2023), color = "#E53935", alpha = 0.6) +
  geom_sf_text(data = cities_sf, aes(label = name), size = 2.5, nudge_y = 0.08) +
  scale_size(range = c(2, 10), name = "Population") +
  theme_void() +
  labs(title = "Albania: Cities (points) on Country (polygon)")
ggsave("output/overlay.png", p_overlay, width = 6, height = 7, dpi = 300)

# ── 5. FILE FORMAT I/O ──────────────────────────────────
# GeoJSON (web-friendly, text)
st_write(cities_sf, "output/cities.geojson", delete_dsn = TRUE)
# GeoPackage (modern binary, replaces Shapefile)
st_write(cities_sf, "output/cities.gpkg", delete_dsn = TRUE)
# Read back
cities_back <- st_read("output/cities.geojson", quiet = TRUE)
cat("\nRead back:", nrow(cities_back), "features\n")

# ── 6. tmap PREVIEW ─────────────────────────────────────
# library(tmap)
# tmap_mode("plot")  # static
# tm_shape(alb_poly) +
#   tm_polygons(col = "#E3F2FD") +
#   tm_shape(cities_sf) +
#   tm_dots(size = "pop_2023", col = "#E53935",
#     title.size = "Population") +
#   tm_text("name", size = 0.7) +
#   tm_layout(title = "Albania")

# tmap_mode("view")  # switch to interactive leaflet!
# Same tm_ code now renders in browser

cat("\n── All W05-M05 R outputs saved ──\n")
